In [ ]:
%matplotlib inline

# Assignment 1

**DUE: Sunday June 28, 2020 11:59pm**

Turn in the assignment via Canvas.

To write legible answers you will need to be familiar with both [Markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet) and [Latex](https://www.latex-tutorial.com/tutorials/amsmath/)

Before you turn this problem in, make sure everything runs as expected. First, restart the kernel (in the menubar, select Kernel→→Restart) and then run all cells (in the menubar, select Cell→→Run All).

Make sure you fill in any place that says "YOUR CODE HERE" or "YOUR ANSWER HERE", as well as your name below:

In [ ]:
NAME = "Juan Lee"
STUDENT_ID = ""

## Problem 1 - Bayes' Theorem

Sadly, many physicians and almost all (non-mathematician) patients would get this problem wrong!

Suppose that a certain form of cancer has a test that correctly detects cancer (if one has it) 99.5% of the time. The prevalence of the cancer in the general population is 1 person in 14,000. Also, if one doesn't have cancer, the test will incorrectly say that one does, $\frac{1}{2}$% of the time (one out of 200).

NOTE: For each of the following questions, please give each answer to at least 2 significant digit accuracy. To accomplish this, don't round off your results until the very last step of each calculation.

### (a) What's the probability that someone picked at random has this form of cancer?

[.000071]


### (b) If you take the test and it comes out positive, what is the probability that you have this form of cancer?

# [.995]



### (c) If you take the test and it comes out negative, what is the probability that you have this form of cancer?

[.005]

## List and Dictionary Comprehensions

In Python, there is a very useful construct called list comprehensions, which in Python 3 also can be used to construct dictionaries.

For one simple example, we can generate a list of random integers using Numpy's randint function.  We can then generate a list of the cubes of those numbers. 

In [ ]:
from numpy.random import randint, seed

# Set the seed so the same random numbers will be generated.
seed(42)

# Create a list of 22 simulated die casts (i.e., 22 random numbers from 1 to 6).
number_list = [randint(1, 7) for _ in range(22)]
print(number_list)

# Now create a list of the cubes of those of the above integers that are less than 5.
cubed = [n**3 for n in number_list if n < 5]
print(cubed)

[4, 5, 3, 5, 5, 2, 3, 3, 3, 5, 4, 3, 6, 5, 2, 4, 6, 6, 2, 4, 5, 1]
[64, 27, 8, 27, 27, 27, 64, 27, 8, 64, 8, 64, 1]


## Problem 2 -  List Comprehensions

### (a) Use list comprehensions to create a list of all the indices of the 2's in the number_list you generated above.

In [ ]:
### YOUR CODE HERE ###
listof2 = []
i = 0
for n in number_list:
  if n == 2:
    listof2.append(i) 
  i+=1
listof2

[5, 14, 18]

### (b) Generate a list of the sum of the numbers that come up when a pair of dice is cast 22 times.

In [ ]:
seed(117)   # DON'T CHANGE

### YOUR CODE HERE ###
random_num = []
sums = []
for i in range(22):
  random_num.append(randint(1, 7))
  sums.append(sum(random_num))
  
sums

[1,
 2,
 4,
 8,
 10,
 16,
 18,
 22,
 23,
 26,
 28,
 32,
 35,
 38,
 42,
 45,
 48,
 54,
 56,
 58,
 61,
 63]

### (c) Generate a list of the sum of the numbers that come up when a pair of dice is cast 100000 times. Find the mean and standard deviation of these numbers.

In [ ]:
from numpy import mean, std

seed(111)   # DON'T CHANGE

### YOUR CODE HERE ###
random_num = []
sums = []
for i in range(100000):
  random_num.append(randint(1, 7))
  sums.append(sum(random_num))

print('mean =', mean(sums), '    std =', std(sums))

mean = 175624.97082     std = 101437.6191798144


## Monte Carlo Simulations

Monte Carlo simulations can yield numeric solutions to probability problems that aren't possible to solve analytically. These simulations are also often easy to code up, and so also provide a way to check one's calculations for problems that amenable to analysis. So this is an important technique to become familiar with.

Basically, one creates a simulation of the situation, and runs many trials that allow one to estimate the probabilities by computing proportions. 

First we give you an example of a Monte Carlo simulation, and then you'll solve such a problem yourself.

### Example

If you throw a die 100 times, with what probability are you likely to see the subsequence {2, 4, 6} appear?

Be patient, this may take a few minutes to run, but notice how straightforward the coding is. The simplicity of the Monte Carlo method makes it very powerful.

In [ ]:
from numpy.random import randint
import numpy as np

seed(17)   # DON'T CHANGE

num_tosses = 100    # tosses per trial
num_trials = 100000

# This sets the feedback interval so we know the program hasn't crashed.
feedback = int(np.round(num_trials / 10))

num_seq_found = 0   # to count the number of target subsequences detected
for t in range(1, num_trials + 1):
    
    # To see the progress.
    if t % feedback == 0:  
        print(np.round(100 * t / num_trials, 1), '%  complete:   prob =', num_seq_found / t)
        
    # Roll the die num_tosses times.
    trial = [randint(1, 7) for _ in range(num_tosses)]
    # Find the indices of all the 2s.
    twos = [i for i in range(num_tosses) if trial[i] == 2]
    
    # Search to see if after each 2 we have a 4 followed by a 6.
    for j in twos:
        if j < num_tosses - 2 and trial[j] == 2 and trial[j + 1] == 4 and trial[j + 2] == 6:
            # We've found the target subsequence.
            num_seq_found += 1
    
print('Probability of this subsequence =', num_seq_found / num_trials)

10.0 %  complete:   prob = 0.4509
20.0 %  complete:   prob = 0.45315
30.0 %  complete:   prob = 0.4488
40.0 %  complete:   prob = 0.448225
50.0 %  complete:   prob = 0.44994
60.0 %  complete:   prob = 0.4504166666666667
70.0 %  complete:   prob = 0.4519
80.0 %  complete:   prob = 0.451175
90.0 %  complete:   prob = 0.4522888888888889
100.0 %  complete:   prob = 0.45287
Probability of this subsequence = 0.45287


This is often a bit of a waiting game, though by exploiting parallelism one can speed it up a bit.

Running the above simulation 5 times previously, we obtained  0.4561, 0.45562, 0.45296, 0.45154, and 0.45529. One can see that the true probability is about 0.454 (their average), though more trials will be needed to obtain a more accurate estimate.  We haven't proved anything, yet we have a very good idea of the true probability. This is a good reality check for many probability calculations. 

The above can easily be calculated precisely. Any particular sequence of three tosses that come up 2,4,6 has probability $\left(\frac{1}{6}\right)^3$, and there are 98 possible starting points for the sequence, so there are 98 ways in which the target sequence could be achieved. So the probability is:

$$\left(\frac{1}{6}\right)^3\cdot 98 = 0.45370370\dots~\dot{=}~0.454$$

One can use Monte Carlo Simulations for far more than just solving probability problems. 

## Problem 3  -  Monte Carlo Simulation

Consider the following game. Randomly pick three balls (without replacement) from an urn containing 60 red balls, 30 green balls, and 10 blue balls. If the majority of balls is red then roll dice and you win an amount equal to the sum of the dice. If the majority of balls isn't red, than you lose the sum of the dice roll. If you had to pay $2 to play this game, about how much would you expect to win (or lose)? Please round off your answer to the nearest cent.

In [ ]:
import numpy as np
from numpy.random import shuffle, randint, seed

seed(213)   # DON'T CHANGE
# The number of times you start with a fresh urn and draw 3 balls.
num_trials = 1000000  # DON'T CHANGE
num_picks = 3
### YOUR CODE HERE ###
trial = []
for i in range(num_trials):
  r = 60
  g = 90
  b = 100
  rcount = 0
  rtwo = False

  temp = randint(1,100)
  if temp <= r:
    rcount+=1
    r-=1
    g-=1
    b-=1
  elif  r< temp <= g:
    g-=1
    b-=1
  elif g < temp <= b:
    b-=1

  temp = randint(1,99)
  if temp <= r:
    rcount+=1
    r-=1
    g-=1
    b-=1
    if rcount == 2:
      rtwo = True
  elif  r< temp <= g:
    g-=1
    b-=1
  elif g < temp <= b:
    b-=1

  temp = randint(1,98)
  if temp <= r:
    rcount+=1
    r-=1
    g-=1
    b-=1
    if rcount == 2:
      rtwo = True
  elif  r< temp <= g:
    g-=1
    b-=1
  elif g < temp <= b:
    b-=1
  temp2 = 0
  if rtwo == True:
    temp2 += randint(1, 7)
    temp2 += randint(1, 7)
    trial.append(-2 + temp2)
    
  else:
    temp2 += randint(1, 7)
    temp2 += randint(1, 7)
    trial.append(-2 - temp2)
    


mean_winnings = mean(trial)

print('mean winnings =', mean_winnings)

mean winnings = 0.223042


[.22]

mean winnings = ?

## Problem 4

Write down the factored conditional probability expression that corresponds
to the graphical Bayesian Network shown bellow:

![bn1](https://alliance.seas.upenn.edu/~cis520/dynamic/2018/wiki/uploads/Lectures/example_trails_bn_small.png)


[p(A, B, C, D, E, F, G, H, I, J) = p(A) p(B|A) p(C|B) p(D|C, E) p(H|A, G) p(I|H) p(G) p(F|G) p(E|F) p(J|G)]




## Problem 5

Given the following Bayesian Network, find the given joint probabilities below. Write out the product of probabilities and conditional probabilites that lead to your numeric answer as well as reporting your numeric answer.

![bn2](https://drive.google.com/uc?id=1TKufu8R4QmgAY2hL3dN8BK6m6jif8L2k)

### (a) $~~~$ $P(Cloudy, Sprinkler, Rain, Wet\_Grass) =$ 



```
`# This is formatted as code`
```

[.0396]

###(b) $~~~$  $P(Cloudy, not Sprinkler, Rain, Wet\_Grass) =$ 


[.324]

### (c)  $~~~$   If 5(a) < 5(b), why?  $~~~$  OR $~~~$  If 5(a) > 5(b), why?




```
# This is formatted as code
```

[5(A) is less than 5(B) because while it was cloudy, having the sprinkler on and having it off makes a huge difference in probability. Having the sprinkler on while it was cloudly multiplies the probability by .1 while having it off would multiply it by .9 instead. p(Cloudy, Rain, Wet_Grass) is the same for part a and b but its the probability of the sprinkler turning off or on that really makes the difference between the two probabilities.]